source/core/slang-short-list.h

```diff
@@ -0,0 +1,487 @@
+            }
+        }
+
+        SLANG_FORCE_INLINE const T& operator [](Index idx) const
+        {
+            SLANG_ASSERT(idx >= 0 && idx < m_count);
+            return (idx < shortListSize) ?
+                m_shortBuffer[idx] : m_buffer[idx - shortListSize];
+        }
+
+        SLANG_FORCE_INLINE T& operator [](Index idx)
+        {
+            SLANG_ASSERT(idx >= 0 && idx < m_count);
+            return (idx < shortListSize) ?
+                m_shortBuffer[idx] : m_buffer[idx - shortListSize];
+        }
+
+        template<typename Func>
+        Index findFirstIndex(const Func& predicate) const
+        {
+            for (Index i = 0; i < Math::Min(m_count, (Index)shortListSize); i++)
+            {
+                if (predicate(m_shortBuffer[i]))
+                    return i;
+            }
+            for (Index i = shortListSize; i < m_count; i++)
+            {
+                if (predicate(m_buffer[i - shortListSize]))
+                    return i;
+            }
+            return -1;
+        }
+
+        template<typename T2>
+        Index indexOf(const T2& val) const
+        {
+            for (Index i = 0; i < Math::Min(m_count, (Index)shortListSize); i++)
+            {
+                if (m_shortBuffer[i] == val)
+                    return i;
+            }
+            for (Index i = shortListSize; i < m_count; i++)
+            {
+                if (m_buffer[i - shortListSize] == val)
+                    return i;
+            }
+            return -1;
+        }
+
+        template<typename Func>
+        Index findLastIndex(const Func& predicate) const
+        {
+            for (Index i = m_count - 1; i >= shortListSize; i--)
+            {
+                if (predicate(m_buffer[i - shortListSize]))
+                    return i;
+            }
+            for (Index i = Math::Min((Index)shortListSize, m_count) - 1; i >= 0; i--)
+            {
+                if (predicate(m_shortBuffer[i]))
+                    return i;
+            }
+            return -1;
+        }
+
+        template<typename T2>
+        Index lastIndexOf(const T2& val) const
+        {
+            for (Index i = m_count - 1; i >= shortListSize; i--)
+            {
+                if (m_buffer[i - shortListSize] == val)
+                    return i;
+            }
+            for (Index i = Math::Min((Index)shortListSize, m_count) - 1; i >= 0; i--)
+            {
+                if (m_shortBuffer[i] == val)
+                    return i;
+            }
+            return -1;
+        }
+
+        bool contains(const T& val) const { return indexOf(val) != Index(-1); }
+
+        template <typename IterateFunc>
+        void forEach(IterateFunc f) const
+        {
+            for (Index i = 0; i < m_count; i++)
+                f(m_buffer[i]);
+        }
+
+    private:
+        T* m_buffer = nullptr;           ///< A new T[N] allocated buffer. NOTE! All elements up to capacity are in some valid form for T.
```

Comment:
Eventually it seems like we will want to change the invariants on `List<T>` and `ShortList<T>` so that they do *not* initialize elements up to the capacity, and instead behave more like standard containers where only the elements used are initialized.